PR: https://github.com/odoo/odoo/pull/164290

From: bb34169dca36370fe6ad370aa97f6e7faa24b0b7
From: Yash Vaishnav
Date: 2024-09-04 11:57:44

Breaking data model changes scores: del:2 + add:0, change matches:
-    request_hour_from = fields.Selection([
-    request_hour_to = fields.Selection([

Total Changes: 63

[IMP] *: improve the custom hour fields in time off

* = hr_holidays, hr_work_entry_holidays

In this commit, the request_hour_from and request_hour_to fields are no
more Selection field
Now it's converted to the Float field which is in HH: MM form

Task-3888748

Part-of: odoo/odoo#164290
Related: odoo/enterprise#61919
Related: odoo/upgrade#5972
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays/models/hr_leave.py
+++ b/addons/hr_holidays/models/hr_leave.py
@@ -204,56 +204,8 @@ class HolidaysRequest(models.Model):
     request_date_from = fields.Date('Request Start Date')
     request_date_to = fields.Date('Request End Date')
     # Interface fields used when using hour-based computation
-    request_hour_from = fields.Selection([
-        ('0', '12:00 AM'), ('0.5', '12:30 AM'),
-        ('1', '1:00 AM'), ('1.5', '1:30 AM'),
-        ('2', '2:00 AM'), ('2.5', '2:30 AM'),
-        ('3', '3:00 AM'), ('3.5', '3:30 AM'),
-        ('4', '4:00 AM'), ('4.5', '4:30 AM'),
-        ('5', '5:00 AM'), ('5.5', '5:30 AM'),
-        ('6', '6:00 AM'), ('6.5', '6:30 AM'),
-        ('7', '7:00 AM'), ('7.5', '7:30 AM'),
-        ('8', '8:00 AM'), ('8.5', '8:30 AM'),
-        ('9', '9:00 AM'), ('9.5', '9:30 AM'),
-        ('10', '10:00 AM'), ('10.5', '10:30 AM'),
-        ('11', '11:00 AM'), ('11.5', '11:30 AM'),
-        ('12', '12:00 PM'), ('12.5', '12:30 PM'),
-        ('13', '1:00 PM'), ('13.5', '1:30 PM'),
-        ('14', '2:00 PM'), ('14.5', '2:30 PM'),
-        ('15', '3:00 PM'), ('15.5', '3:30 PM'),
-        ('16', '4:00 PM'), ('16.5', '4:30 PM'),
-        ('17', '5:00 PM'), ('17.5', '5:30 PM'),
-        ('18', '6:00 PM'), ('18.5', '6:30 PM'),
-        ('19', '7:00 PM'), ('19.5', '7:30 PM'),
-        ('20', '8:00 PM'), ('20.5', '8:30 PM'),
-        ('21', '9:00 PM'), ('21.5', '9:30 PM'),
-        ('22', '10:00 PM'), ('22.5', '10:30 PM'),
-        ('23', '11:00 PM'), ('23.5', '11:30 PM')], string='Hour from')
-    request_hour_to = fields.Selection([
-        ('0', '12:00 AM'), ('0.5', '12:30 AM'),
-        ('1', '1:00 AM'), ('1.5', '1:30 AM'),
-        ('2', '2:00 AM'), ('2.5', '2:30 AM'),
-        ('3', '3:00 AM'), ('3.5', '3:30 AM'),
-        ('4', '4:00 AM'), ('4.5', '4:30 AM'),
-        ('5', '5:00 AM'), ('5.5', '5:30 AM'),
-        ('6', '6:00 AM'), ('6.5', '6:30 AM'),
-        ('7', '7:00 AM'), ('7.5', '7:30 AM'),
-        ('8', '8:00 AM'), ('8.5', '8:30 AM'),
-        ('9', '9:00 AM'), ('9.5', '9:30 AM'),
-        ('10', '10:00 AM'), ('10.5', '10:30 AM'),
-        ('11', '11:00 AM'), ('11.5', '11:30 AM'),
-        ('12', '12:00 PM'), ('12.5', '12:30 PM'),
-        ('13', '1:00 PM'), ('13.5', '1:30 PM'),
-        ('14', '2:00 PM'), ('14.5', '2:30 PM'),
-        ('15', '3:00 PM'), ('15.5', '3:30 PM'),
-        ('16', '4:00 PM'), ('16.5', '4:30 PM'),
-        ('17', '5:00 PM'), ('17.5', '5:30 PM'),
-        ('18', '6:00 PM'), ('18.5', '6:30 PM'),
-        ('19', '7:00 PM'), ('19.5', '7:30 PM'),
-        ('20', '8:00 PM'), ('20.5', '8:30 PM'),
-        ('21', '9:00 PM'), ('21.5', '9:30 PM'),
-        ('22', '10:00 PM'), ('22.5', '10:30 PM'),
-        ('23', '11:00 PM'), ('23.5', '11:30 PM')], string='Hour to')
+    request_hour_from = fields.Float(string='Hour from')
+    request_hour_to = fields.Float(string='Hour to')
     # used only when the leave is taken in half days
     request_date_from_period = fields.Selection([
         ('am', 'Morning'), ('pm', 'Afternoon')],
@@ -279,6 +231,17 @@ class HolidaysRequest(models.Model):
                            self._table, ['date_to', 'date_from'])
         return res
 
+    @api.onchange('request_hour_from', 'request_hour_to')
+    def _onchange_hours(self):
+        # avoid negative or after midnight
+        self.request_hour_from = min(self.request_hour_from, 23.99)
+        self.request_hour_from = max(self.request_hour_from, 0.0)
+        self.request_hour_to = min(self.request_hour_to, 24)
+        self.request_hour_to = max(self.request_hour_to, 0.0)
+
+        # avoid wrong order
+        self.request_hour_to = max(self.request_hour_to, self.request_hour_from)
+
     @api.depends_context('uid')
     def _compute_description(self):
         self.check_access_rights('read')
